### PR TITLE
Uniqueness checks of provider keys and addresses in ForcePayment

### PR DESCRIPTION
**Pull request by Jakub89**
Resolves: #664 
Resolves: #665

### CODE REVIEW

File: concent_api/core/subtask_helpers.py [hunk 1]
```diff
@@ -150,12 +152,36 @@ def update_subtask_state(
...
+    is_requestor_public_key_unique = (len(unique_requestor_public_keys) == 1)
+    is_requestor_ethereum_address_unique = (len(unique_requestor_ethereum_addresses) == 1)
+    is_requestor_ethereum_public_key_unique = (len(unique_requestor_ethereum_public_keys) == 1)
+
+    is_provider_public_key_unique = (len(unique_provider_public_keys) == 1)
+    is_provider_ethereum_address_unique = (len(unique_provider_ethereum_addresses) == 1)
+    is_provider_ethereum_public_key_unique = (len(unique_provider_ethereum_public_keys) == 1)
+
+    requestor_public_key = subtask_results_accepted_list[0].task_to_compute.requestor_public_key
+    unique_signature = all(
```
Comment (dybi) [inline concent_api/core/subtask_helpers.py]: I would suggest renaming it to: `are_all_messages_signed_by_requestor` or shorter `are_all_signed_by_requestor`, or even shorter `all_signed_by_requestor`

File: concent_api/core/subtask_helpers.py [hunk 2]
```diff
@@ -150,12 +152,36 @@ def update_subtask_state(
...
-    return is_public_key_unique and is_ethereum_address_unique and is_ethereum_public_key_unique
+    is_requestor_public_key_unique = (len(unique_requestor_public_keys) == 1)
+    is_requestor_ethereum_address_unique = (len(unique_requestor_ethereum_addresses) == 1)
+    is_requestor_ethereum_public_key_unique = (len(unique_requestor_ethereum_public_keys) == 1)
+
+    is_provider_public_key_unique = (len(unique_provider_public_keys) == 1)
+    is_provider_ethereum_address_unique = (len(unique_provider_ethereum_addresses) == 1)
+    is_provider_ethereum_public_key_unique = (len(unique_provider_ethereum_public_keys) == 1)
+
+    requestor_public_key = subtask_results_accepted_list[0].task_to_compute.requestor_public_key
```
Comment (dybi) [inline concent_api/core/subtask_helpers.py]: as new functionality has been added, some more tests are needed

File: concent_api/core/subtask_helpers.py [hunk 3]
```diff
@@ -150,12 +152,36 @@ def update_subtask_state(
...
 
-    is_public_key_unique = (len(unique_public_keys) == 1)
-    is_ethereum_address_unique = (len(unique_ethereum_addresses) == 1)
-    is_ethereum_public_key_unique = (len(unique_ethereum_public_keys) == 1)
+    unique_provider_public_keys = set(subtask_results_accepted.task_to_compute.provider_public_key for subtask_results_accepted in subtask_results_accepted_list)
+    unique_provider_ethereum_addresses = set(subtask_results_accepted.task_to_compute.provider_ethereum_address for subtask_results_accepted in subtask_results_accepted_list)
+    unique_provider_ethereum_public_keys = set(subtask_results_accepted.task_to_compute.provider_ethereum_public_key for subtask_results_accepted in subtask_results_accepted_list)
 
-    return is_public_key_unique and is_ethereum_address_unique and is_ethereum_public_key_unique
+    is_requestor_public_key_unique = (len(unique_requestor_public_keys) == 1)
```
Comment (dybi) [inline concent_api/core/subtask_helpers.py]: as new functionality has been added, some more tests are needed

File: concent_api/core/subtask_helpers.py (hunk 1 above)
Comment (Jakub89) [reply concent_api/core/subtask_helpers.py]: Done

File: concent_api/core/subtask_helpers.py (hunk 2 above)
Comment (Jakub89) [reply concent_api/core/subtask_helpers.py]: Done

File: concent_api/core/subtask_helpers.py (hunk 3 above)
Comment (Jakub89) [reply concent_api/core/subtask_helpers.py]: Done

File: concent_api/core/subtask_helpers.py [hunk 4]
```diff
@@ -150,12 +152,37 @@ def update_subtask_state(
...
-    return is_public_key_unique and is_ethereum_address_unique and is_ethereum_public_key_unique
+def are_subtask_results_accepted_messages_sign_by_the_same_requestor(subtask_results_accepted_list: List[SubtaskResultsAccepted]) -> bool:
+    requestor_public_key = subtask_results_accepted_list[0].task_to_compute.requestor_public_key
+    are_all_signed_by_requestor = all(
+        is_golem_message_signed_with_key(
+            hex_to_bytes_convert(requestor_public_key),
+            subtask_results_accepted
+        ) for subtask_results_accepted in subtask_results_accepted_list
+    )
+    return are_all_signed_by_requestor
```
Comment (dybi) [inline concent_api/core/subtask_helpers.py]: Please add a new line ;)

File: concent_api/core/subtask_helpers.py [hunk 5]
```diff
@@ -150,12 +152,37 @@ def update_subtask_state(
...
+        is_requestor_public_key_unique and
+        is_requestor_ethereum_address_unique and
+        is_requestor_ethereum_public_key_unique and
+        is_provider_public_key_unique and
+        is_provider_ethereum_address_unique and
+        is_provider_ethereum_public_key_unique
+    )
 
-    return is_public_key_unique and is_ethereum_address_unique and is_ethereum_public_key_unique
+def are_subtask_results_accepted_messages_sign_by_the_same_requestor(subtask_results_accepted_list: List[SubtaskResultsAccepted]) -> bool:
```
Comment (dybi) [inline concent_api/core/subtask_helpers.py]: sign -> signed

Comment (dybi) [inline concent_api/core/subtask_helpers.py]: At lease one unit test would be nice ;)

File: concent_api/core/subtask_helpers.py [hunk 6]
```diff
@@ -168,20 +168,21 @@ def are_keys_and_addresses_unique_in_message_subtask_results_accepted(subtask_re
...
         is_requestor_public_key_unique and
         is_requestor_ethereum_address_unique and
         is_requestor_ethereum_public_key_unique and
         is_provider_public_key_unique and
         is_provider_ethereum_address_unique and
         is_provider_ethereum_public_key_unique
     )
+
+def are_subtask_results_accepted_messages_sign_by_the_same_requestor(subtask_results_accepted_list: List[SubtaskResultsAccepted]) -> bool:
+    requestor_public_key = subtask_results_accepted_list[0].task_to_compute.requestor_public_key
```
Comment (rwrzesien) [inline concent_api/core/subtask_helpers.py]: Shouldn't you check if list has at least one element ?

File: concent_api/core/subtask_helpers.py [hunk 7]
```diff
@@ -168,20 +168,21 @@ def are_keys_and_addresses_unique_in_message_subtask_results_accepted(subtask_re
...
+    are_all_signed_by_requestor = all(
+        is_golem_message_signed_with_key(
+            hex_to_bytes_convert(requestor_public_key),
+            subtask_results_accepted
+        ) for subtask_results_accepted in subtask_results_accepted_list
+    )
+    return are_all_signed_by_requestor
```
Comment (rwrzesien) [inline concent_api/core/subtask_helpers.py]: No newline at the end of the file.

File: concent_api/core/subtask_helpers.py (hunk 4 above)
Comment (Jakub89) [reply concent_api/core/subtask_helpers.py]: God damn pyCharm

File: concent_api/core/subtask_helpers.py (hunk 6 above)
Comment (Jakub89) [reply concent_api/core/subtask_helpers.py]: We have that check elsewhere. Look at the `test_provider_send_force_payment_with_empty_subtask_results_accepted_list_concent_should_return_http_400` in `core/tests/test_integration_force_payment.py` it returns 
`b'{"error": "subtask_results_accepted_list must be a list type and contains at least one message", "error_code": "message.value_wrong_length"}'`

**Comment by Jakub89** [comment]
Merged

**Pull request closed without merging**